lib/marin/src/marin/processing/tokenize/tokenize.py

```diff
@@ -384,24 +253,150 @@ def _get_filepaths_to_tokenize(input_paths: list[str]) -> list[str]:
     return out
 
 
-def _is_probably_path(path: str) -> bool:
-    """see if it looks like a real path or not, in which case it might be an hf dataset"""
+def _group_files_by_size(file_infos, max_bytes: int):
+    """Group files into lists by total size threshold."""
+    current_group = []
+    current_size = 0
+
+    for info in file_infos:
+        if current_size + info["size"] >= max_bytes and current_group:
+            yield current_group
+            current_group = []
+            current_size = 0
+        current_group.append(info["filename"])
+        current_size += info["size"]
+
+    if current_group:
+        yield current_group
+
+
+def _tokenize_batches(config: TokenizeConfig | HfTokenizeConfig, batches):
+    """Load tokenizer once per shard and process all batches."""
+    tokenizer = transformers.AutoTokenizer.from_pretrained(config.tokenizer)
+    batch_processor = preprocessor_for_format(config.format, tokenizer)
+
+    for batch in batches:
+        yield from batch_processor(batch)
+
+
+def _resolve_hf_revision(dataset_id: str, revision: str | None) -> str:
+    """Resolve HuggingFace dataset revision using the HuggingFace Hub API."""
+    if revision is not None:
+        return revision
+
+    info = dataset_info(dataset_id)
+    assert info.sha is not None, "Failed to resolve dataset revision from HuggingFace Hub."
+    return info.sha
+
+
+def tokenize(config: TokenizeConfigBase):
+    """Tokenize datasets using zephyr pipeline.
+
+    Processes train and validation splits separately, writing to Levanter cache format.
+    For HuggingFace datasets, downloads them first then tokenizes the downloaded files.
+    """
+
+    if isinstance(config, TokenizeConfig):
+        train_paths = _get_filepaths_to_tokenize(config.train_paths) if config.train_paths else []
+        validation_paths = _get_filepaths_to_tokenize(config.validation_paths) if config.validation_paths else []
+    elif isinstance(config, HfTokenizeConfig):
+        # Use HfFileSystem to construct hf:// URLs for streaming without downloading
+        try:
+            from huggingface_hub import HfFileSystem
+        except ImportError as e:
+            raise ImportError(
+                "HfTokenizeConfig requires huggingface_hub. Install it with: uv pip install huggingface_hub"
+            ) from e
+
+        hf_fs = HfFileSystem(token=os.environ.get("HF_TOKEN", False))
```

Comment:
technically we lose our ability to handle custom datasets but i think they are moving away from that